File: c++/SwapNodePairs.cpp
class Solution {
	public:
		ListNode* reverse(ListNode* curNode) {
			if (curNode == NULL) {
				return curNode;
			}

			ListNode* first = curNode;
			ListNode* second = first->next;

			if (second) {
				first->next = reverse(second->next);
				second->next = first;
				return second;
			} else {
				return first;
			}
		}

		ListNode* swapPairs(ListNode* head) {
			if (head == NULL || head->next == NULL) {
				return head;
			}

			return reverse(head);
		}
};

File: python3/SwapNodePairs.py
class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        if head == None or head.next == None:
            return head
        
        def reverse(curNode):
            if curNode == None:
                return None
            
            first = curNode
            second = curNode.next
            
            if second:
                first.next = reverse(second.next)
                second.next = first
                return second
            else:
                return first

        return reverse(head)
